### PR TITLE
Support passing routes as well as worker names to `wrangler tail`

### PR DESCRIPTION
**Pull request by penalosa**
An initial stab at https://github.com/cloudflare/wrangler2/issues/492, following the approach outlined there:
> 1. Checking if the positional argument passed to wrangler tail <url | name> resembles a URL
> 1. Determining the zone ID from the URL (e.g. via [getZoneForRoute](https://github.com/cloudflare/wrangler2/blob/main/packages/wrangler/src/zones.ts#L29))
> 1. Calling [GET zone/:zone_id/workers/routes](https://api.cloudflare.com/#worker-routes-list-routes)
> 1. Finding the given URL in the returned routes
> 1. Using the worker name associated with that route to plug in to the existing tail infrastructure

Some notes:
- Distinguishing between a route and a worker name is done by checking for the presence of a `.` character, following confirmation that that will suffice.
- I've defined a new type for responses from the route list endpoint. I couldn't find anything similar, but I may have missed an existing type that should be used instead:
```
interface WorkerRoute {
	id: string;
	pattern: string;
	script: string;
}
```
- I've tested this on a worker in my CF account, but I'm unsure as to how to go about adding automated tests for the new functionality—some pointers here would be really helpful!

This PR covers exact route matching, whereas the linked issue asks for flexible route matching. Some thoughts on that:
- **Precedence** For instance, if two routes were defined (`www.example.com` and `*.example.com`), and the logs for `www.example.com` were requested, Wrangler would have to figure out that the worker on the `www.example.com` route would have served those requests, even though the `*.example.com` route would also match. This is obviously implemented somewhere, and so the precedence algorithm could be reimplemented in Wrangler to allow for this (maybe it's already implemented somewhere in Wrangler?)
- **Matching rules** The matching rules for worker routes are [documented](https://developers.cloudflare.com/workers/platform/routing/routes/#matching-behavior) so they could be implemented in Wrangler. This would be for e.g. `wrangler tail example.com/foo` to find a worker deployed on the `example.com/*` route. 
- **Filtering** A related concern is filtering URL patterns. Currently, the logs backend only supports filtering based on IP, header etc... and not on URL path. That filtering could either be done within Wrangler, or added as a filter to the logs backend, to make sure that e.g. `wrangler tail example.com/foo`, matching a worker deployed on the route `example.com/*`, only showed the logs from requests to `example.com/foo`, and not requests to `example.com/baz` (which would be handled by the same worker).

### CODE REVIEW

File: .changeset/ninety-books-hope.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"wrangler": patch
+---
+
+Closes [#1505](https://github.com/cloudflare/wrangler2/issues/1505) by extending `wrangler tail` to allow for passing worker routes as well as worker script names.
```
Comment (caass) [inline .changeset/ninety-books-hope.md:5]: Maybe an example? Something like "`wrangler tail example.com/*` will work if you have a worker assigned to the route example.com/*`

File: packages/wrangler/src/index.tsx [hunk 2]
```diff
@@ -609,7 +610,7 @@ function createCLIParser(argv: string[]) {
 		(yargs) => {
 			return yargs
 				.positional("name", {
-					describe: "Name of the worker",
+					describe: "Name of the worker, or route at which it runs",
```
Comment (caass) [inline packages/wrangler/src/index.tsx:613]:
```js
.positional("worker", {
  describe: "Name or route of the worker to tail"
```

File: packages/wrangler/src/index.tsx [hunk 3]
```diff
@@ -677,7 +678,34 @@ function createCLIParser(argv: string[]) {
...
 			});
 
-			const scriptName = getLegacyScriptName(args, config);
+			let scriptName;
+
+			// Worker names can't contain "." (and most routes should), so use that as a discriminator
+			if (args.name?.includes(".")) {
+				const zone = await getZoneForRoute(args.name);
+				if (!zone) {
+					throw new Error(`Could not find zone for route ${args.name}`);
```
Comment (caass) [inline packages/wrangler/src/index.tsx:687]: This error message should be more actionable by the user -- describe what exactly went wrong and how they can fix it.

File: packages/wrangler/src/index.tsx [hunk 4]
```diff
@@ -677,7 +678,34 @@ function createCLIParser(argv: string[]) {
...
+							logger.log(
+								`Connecting to worker ${scriptName} at route ${args.name}`
+							);
+						}
+						break;
+					}
+				}
+				if (!scriptName) {
+					throw new Error(`No workers found on the route ${args.name}`);
+				}
```
Comment (caass) [inline packages/wrangler/src/index.tsx:705]: Maybe refactor into a `getWorkerForRoute` function?

File: packages/wrangler/src/index.tsx [hunk 5]
```diff
@@ -677,7 +678,34 @@ function createCLIParser(argv: string[]) {
...
+					if (route.pattern === args.name) {
+						scriptName = route.script;
+						if (args.format === "pretty") {
+							logger.log(
+								`Connecting to worker ${scriptName} at route ${args.name}`
+							);
+						}
+						break;
+					}
+				}
```
Comment (caass) [inline packages/wrangler/src/index.tsx:702]: `routes.find(...`

File: packages/wrangler/src/index.tsx [hunk 6]
```diff
@@ -677,7 +678,34 @@ function createCLIParser(argv: string[]) {
...
+				const routes = await getRoutesForZone(zone.id);
+
+				// Find an exact match for the route provided
+				for (const route of routes) {
+					if (route.pattern === args.name) {
+						scriptName = route.script;
+						if (args.format === "pretty") {
+							logger.log(
+								`Connecting to worker ${scriptName} at route ${args.name}`
```
Comment (caass) [inline packages/wrangler/src/index.tsx:697]: does tail already say what worker we're connecting to?

Comment (caass) [reply packages/wrangler/src/index.tsx:697]: Yes, but later in the process -- if the connection fails, we won't know what we're connecting to.

Comment (caass) [reply packages/wrangler/src/index.tsx:697]: So maybe it's better to add a logger statement earlier in the tail lifecycle, before anything can fail?

File: packages/wrangler/src/index.tsx [hunk 7]
```diff
@@ -677,7 +678,34 @@ function createCLIParser(argv: string[]) {
...
+							);
+						}
+						break;
+					}
+				}
+				if (!scriptName) {
+					throw new Error(`No workers found on the route ${args.name}`);
```
Comment (caass) [inline packages/wrangler/src/index.tsx:704]: Let's improve this error message by suggesting routes they might have meant.

**Review by caass — CHANGES_REQUESTED** [review]
This looks really good! This will be a great feature and this implementation is on the right track. I left some comments, some super minor, some requiring a little refactoring.

Also, please put doc comments on any functions and/or types so that someone else can tell at a glance what their purpose is :)

**Comment by caass** [comment]
> * **Precedence** For instance, if two routes were defined (`www.example.com` and `*.example.com`), and the logs for `www.example.com` were requested, Wrangler would have to figure out that the worker on the `www.example.com` route would have served those requests, even though the `*.example.com` route would also match. This is obviously implemented somewhere, and so the precedence algorithm could be reimplemented in Wrangler to allow for this (maybe it's already implemented somewhere in Wrangler?)

At the moment, if I understand correctly, this isn't an issue because we require an exact match on routes. But I see what you're saying.

> * **Matching rules** The matching rules for worker routes are [documented](https://developers.cloudflare.com/workers/platform/routing/routes/#matching-behavior) so they could be implemented in Wrangler. This would be for e.g. `wrangler tail example.com/foo` to find a worker deployed on the `example.com/*` route.

If this is documented, we could actually re-implement this in wrangler fairly easily, unless I'm mistaken. We should totally do this!

> * **Filtering** A related concern is filtering URL patterns. Currently, the logs backend only supports filtering based on IP, header etc... and not on URL path. That filtering could either be done within Wrangler, or added as a filter to the logs backend, to make sure that e.g. `wrangler tail example.com/foo`, matching a worker deployed on the route `example.com/*`, only showed the logs from requests to `example.com/foo`, and not requests to `example.com/baz` (which would be handled by the same worker).

So this is actually two issues:

1. If the user passes a URL that is _more specific_ than the matching route (and therefore worker), should we filter logs to only print those that occur at that given URL?
2. Should we implement a filter in the trace worker that allows for filtering of `FetchEvent`s based on the URL they hit?

I think the answer to 1 is no, but the answer to 2 is kind of unclear -- and also outside the scope of this PR. It could be worth seeing if this is a feature people want, and if so we could implement it. But so far, I haven't really seen any requests for this.

Some other thoughts:
- What happens if multiple workers would be triggered on a given route? e.g. a worker is assigned to `example.com/*,` and another worker is assigned to `example.com/api/*`, and the URL given is `example.com/api/v1`?
- Are we making it clear enough exactly what worker and what route we are connecting to? Because we may match a route slightly different from what the user inputs (when we implement the lookup rules as documented in the matching rules section), it should be extremely obvious to the user what exactly they are getting logs from.

**Comment by penalosa** [comment]
> What happens if multiple workers would be triggered on a given route? e.g. a worker is assigned to example.com/*, and another worker is assigned to example.com/api/*, and the URL given is example.com/api/v1?

I _think_ only one of those workers would run, governed by the precedence rules (which is why we'd need to implement those, so wrangler could figure out that tailing the `example.com/api/*` worker is the right thing to do). 

From https://developers.cloudflare.com/workers/platform/routing/routes/#matching-behavior

> When more than one route pattern could match a request URL, the most specific route pattern wins. For example, the pattern `www.example.com/*` would take precedence over `*.example.com/*` when matching a request for `https://www.example.com/`. The pattern `example.com/hello/*` would take precedence over `example.com/*` when matching a request for `example.com/hello/world`

File: packages/wrangler/src/zones.ts [hunk 8]
```diff
@@ -74,3 +74,94 @@ export async function getZoneIdFromHost(host: string): Promise<string> {
...
+				distanceBetween(a, b.slice(1), cache),
+				distanceBetween(a.slice(1), b.slice(1), cache)
+			);
+	}
+	cache.set(`${a}|${b}`, result);
+	return result;
+}
+
+/**
+ * Given a route which isn't valid, sort the valid routes by closeness (levenstein distance)
```
Comment (rozenmd) [inline packages/wrangler/src/zones.ts:125]: total nitpick but:

```suggestion
 * Given an invalid route, sort the valid routes by closeness to the invalid route (levenstein distance)
```

**Review by caass — APPROVED** [review]
Awesome! Looks good to me 👍🏻

**Pull request merged**